packages/puppeteer-core/src/common/NetworkManager.ts

```diff
@@ -68,102 +66,90 @@ export const NetworkManagerEmittedEvents = {
+
+  #handlers = new Map<string, Function>([
+    ['Fetch.requestPaused', this.#onRequestPaused],
+    ['Fetch.authRequired', this.#onAuthRequired],
+    ['Network.requestWillBeSent', this.#onRequestWillBeSent],
+    ['Network.requestServedFromCache', this.#onRequestServedFromCache],
+    ['Network.responseReceived', this.#onResponseReceived],
+    ['Network.loadingFinished', this.#onLoadingFinished],
+    ['Network.loadingFailed', this.#onLoadingFailed],
+    ['Network.responseReceivedExtraInfo', this.#onResponseReceivedExtraInfo],
   ]);
 
-  constructor(
-    client: CDPSession,
-    ignoreHTTPSErrors: boolean,
-    frameManager: Pick<FrameManager, 'frame'>
-  ) {
+  #clients = new Set<CDPSession>();
+
+  constructor(ignoreHTTPSErrors: boolean, frameManager: FrameProvider) {
     super();
-    this.#client = client;
     this.#ignoreHTTPSErrors = ignoreHTTPSErrors;
     this.#frameManager = frameManager;
-
-    for (const [event, handler] of this.#handlers) {
-      this.#client.on(event, handler);
-    }
   }
 
-  async updateClient(client: CDPSession): Promise<void> {
-    this.#client = client;
-    for (const [event, handler] of this.#handlers) {
-      this.#client.on(event, handler);
+  async addClient(client: CDPSession): Promise<void> {
+    if (this.#clients.has(client)) {
+      return;
     }
-    this.#deferredInit = undefined;
-    await this.initialize();
-  }
-
-  /**
-   * Initialize calls should avoid async dependencies between CDP calls as those
-   * might not resolve until after the target is resumed causing a deadlock.
-   */
-  initialize(): Promise<void> {
-    if (this.#deferredInit) {
-      return this.#deferredInit.valueOrThrow();
+    this.#clients.add(client);
+    for (const [event, handler] of this.#handlers) {
+      client.on(event, handler.bind(this, client));
     }
-    this.#deferredInit = createDebuggableDeferred(
-      'NetworkManager initialization timed out'
+    client.on(
+      CDPSessionEmittedEvents.Disconnected,
+      this.#removeClient.bind(this, client)
     );
-    const init = Promise.all([
+    await Promise.all([
       this.#ignoreHTTPSErrors
-        ? this.#client.send('Security.setIgnoreCertificateErrors', {
+        ? client.send('Security.setIgnoreCertificateErrors', {
             ignore: true,
           })
         : null,
-      this.#client.send('Network.enable'),
+      client.send('Network.enable'),
+      this.#applyExtraHTTPHeaders(client),
+      this.#applyNetworkConditions(client),
+      this.#applyProtocolCacheDisabled(client),
+      this.#applyProtocolRequestInterception(client),
+      this.#applyUserAgent(client),
     ]);
-    const deferredInitPromise = this.#deferredInit;
-    init
-      .then(() => {
-        deferredInitPromise.resolve();
-      })
-      .catch(err => {
-        deferredInitPromise.reject(err);
-      });
-    return this.#deferredInit.valueOrThrow();
+  }
+
+  async #removeClient(client: CDPSession) {
+    this.#clients.delete(client);
```

Comment:
Remove event listeners from client?

Reply:
Not needed because the client itself is disposed.

Reply:
Can you point me to the code here I see no clearing of the attached events.

Reply:
When nothing references the client, it will be gc’d, its listeners won’t be referenced by anything in turn and will be cleaned up.

Reply:
I will add explicit event listener tracking in the next commit here.